mllib/src/main/scala/org/apache/spark/ml/stat/ANOVATest.scala

```diff
@@ -0,0 +1,167 @@
+
+    SchemaUtils.checkColumnType(dataset.schema, featuresCol, new VectorUDT)
+    SchemaUtils.checkNumericType(dataset.schema, labelCol)
+
+    val labeledPointRdd = dataset.select(col("label").cast("double"), col("features"))
+      .as[(Double, Vector)]
+      .rdd.map { case (label, features) => LabeledPoint(label, features) }
+
+    val numFeatures = labeledPointRdd.first().features.size
+    val numSamples = labeledPointRdd.count()
+    val numClasses = labeledPointRdd.map(d => d.label).distinct.count
+
+    labeledPointRdd.flatMap { case LabeledPoint(label, features) =>
+      features.iterator.map { case (col, value) =>
+        (col, (label, value, value * value))
+      }
+    }.aggregateByKey[(Double, Double, OpenHashMap[Double, Double], OpenHashMap[Double, Long])](
+      (0.0, 0.0, new OpenHashMap[Double, Double], new OpenHashMap[Double, Long]))(
+      seqOp = {
+        case (
+          (sum: Double, sumOfSq: Double, mapOfSumPerClass, mapOfCountPerClass),
+          (label, feature, featureSq)
+         ) =>
+          mapOfSumPerClass.changeValue(label, feature, _ + feature)
+          mapOfCountPerClass.changeValue(label, 1L, _ + 1L)
+          (mapOfSumPerClass, mapOfCountPerClass)
+          (sum + feature, sumOfSq + featureSq, mapOfSumPerClass, mapOfCountPerClass)
+      },
+      combOp = {
+        case (
+          (sum1, sumOfSq1, mapOfSumPerClass1, mapOfCountPerClass1),
+          (sum2, sumOfSq2, mapOfSumPerClass2, mapOfCountPerClass2)
+        ) =>
+          mapOfSumPerClass2.foreach { case (v, w) =>
+            mapOfSumPerClass1.changeValue(v, w, _ + w)
+          }
+          mapOfCountPerClass2.foreach { case (v, w) =>
+            mapOfCountPerClass1.changeValue(v, w, _ + w)
+          }
+          (sum1 + sum2, sumOfSq1 + sumOfSq2, mapOfSumPerClass1, mapOfCountPerClass1)
+      }
+    ).map {
+      case (col, (sum, sumOfSq, mapOfSumPerClass, mapOfCountPerClass)) =>
+        // e.g. feature is [3.3, 2.5, 1.0, 3.0, 2.0] and labels are [1, 2, 1, 3, 3]
+        // sum: sum of all the features (3.3+2.5+1.0+3.0+2.0)
+        // sumOfSq: sum of squares of all the features (3.3^2+2.5^2+1.0^2+3.0^2+2.0^2)
+        // mapOfSumPerClass key: label, value: sum of features for each label
+        //                                         ( 1 -> 3.3 + 1.0, 2 -> 2.5, 3 -> 3.0 + 2.0 )
+        // mapOfCountPerClass key: label, value: count of features for each label
+        //                                         ( 1 -> 2, 2 -> 2, 3 -> 2 )
+        // sqSum: square of sum of all data ((3.3+2.5+1.0+3.0+2.0)^2)
+        val sqSum = sum * sum
+        val ssTot = sumOfSq - sqSum / numSamples
+
+        // sumOfSqSumPerClass:
+        //     sum( sq_sum_classes[k] / n_samples_per_class[k] for k in range(n_classes))
+        //     e.g. ((3.3+1.0)^2 / 2 + 2.5^2 / 1 + (3.0+2.0)^2 / 2)
+        var sumOfSqSumPerClass = 0.0
+        val (keys1, values1) = mapOfSumPerClass.iterator.toSeq.sortBy(_._1).unzip
```

Comment:
is .iterator needed or can you do straight to .toSeq ?

Reply:
Removed ```.iterator```